app/src/main/java/org/thoughtcrime/securesms/components/mention/MentionRenderer.java

```diff
@@ -0,0 +1,133 @@
+      int lineTop    = getLineTop(layout, startLine);
+      int lineBottom = getLineBottom(layout, startLine);
+      int left       = Math.min(startOffset, endOffset);
+      int right      = Math.max(startOffset, endOffset);
+
+      drawable.setBounds(left, lineTop, right, lineBottom);
+      drawable.draw(canvas);
+    }
+  }
+
+  public static final class MultiLineMentionRenderer extends MentionRenderer {
+
+    private final Drawable drawableLeft;
+    private final Drawable drawableMid;
+    private final Drawable drawableRight;
+
+    public MultiLineMentionRenderer(int horizontalPadding, int verticalPadding,
+                                    @NonNull Drawable drawableLeft,
+                                    @NonNull Drawable drawableMid,
+                                    @NonNull Drawable drawableRight)
+    {
+      super(horizontalPadding, verticalPadding);
+      this.drawableLeft  = drawableLeft;
+      this.drawableMid   = drawableMid;
+      this.drawableRight = drawableRight;
+    }
+
+    @Override
+    public void draw(@NonNull Canvas canvas, @NonNull Layout layout, int startLine, int endLine, int startOffset, int endOffset) {
+      int paragraphDirection = layout.getParagraphDirection(startLine);
+
+      float lineEndOffset;
+      if (paragraphDirection == Layout.DIR_RIGHT_TO_LEFT) {
+        lineEndOffset = layout.getLineLeft(startLine) - horizontalPadding;
+      } else {
+        lineEndOffset = layout.getLineRight(startLine) + horizontalPadding;
+      }
+
+      int lineBottom = getLineBottom(layout, startLine);
+      int lineTop    = getLineTop(layout, startLine);
+      drawStart(canvas, startOffset, lineTop, (int) lineEndOffset, lineBottom);
+
+      for (int line = startLine + 1; line < endLine; line++) {
+        int left  = (int) layout.getLineLeft(line) - horizontalPadding;
+        int right = (int) layout.getLineRight(line) + horizontalPadding;
+
+        lineTop    = getLineTop(layout, line);
+        lineBottom = getLineBottom(layout, line);
+
+        drawableMid.setBounds(left, lineTop, right, lineBottom);
+        drawableMid.draw(canvas);
+      }
+
+      float lineStartOffset;
+      if (paragraphDirection == Layout.DIR_RIGHT_TO_LEFT) {
+        lineStartOffset = layout.getLineRight(startLine) + horizontalPadding;
+      } else {
+        lineStartOffset = layout.getLineLeft(startLine) - horizontalPadding;
+      }
+
+      lineBottom = getLineBottom(layout, endLine);
+      lineTop    = getLineTop(layout, endLine);
+
+      drawEnd(canvas, (int) lineStartOffset, lineTop, endOffset, lineBottom);
+    }
+
+    private void drawStart(@NonNull Canvas canvas, int start, int top, int end, int bottom) {
+      if (start > end) {
+        drawableRight.setBounds(end, top, start, bottom);
+        drawableRight.draw(canvas);
+      } else {
+        drawableLeft.setBounds(start, top, end, bottom);
+        drawableLeft.draw(canvas);
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 114:120, 124:130.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.